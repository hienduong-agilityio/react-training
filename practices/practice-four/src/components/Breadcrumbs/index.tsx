// Libraries
import { Link } from 'react-router-dom';
import { memo } from 'react';

// SVG
import homeIcon from '@public/images/homeIcon.svg';
import arrowRight from '@public/images/arrowRightIcon.svg';

// Constants
import { ROUTE } from '@/constants';

export type BreadcrumbItem = {
  // id: The breadcrumb Items id
  id?: number;
  // label: The label to display for the breadcrumb item
  label?: string;
  // href: Optional URL for the breadcrumb item
  href?: string;
};

interface IBreadcrumbsProps {
  // items: Array of breadcrumb items
  items?: BreadcrumbItem[];
}

/**
 * Breadcrumbs component
 *
 * @returns {JSX.Element} Breadcrumbs element
 */
const Breadcrumbs = ({ items = [{ label: 'Home', href: ROUTE.ROOT }] }: IBreadcrumbsProps): JSX.Element => {
  return (
    <nav aria-label='breadcrumb'>
      <ul className='flex items-center space-x-1 md:space-x-2 rtl:space-x-reverse'>
        <li className='inline-flex items-center'>
          <Link to={ROUTE.ROOT} className='flex items-center text-sm font-medium text-gray-400 hover:text-primary-500'>
            <img src={homeIcon} alt='home' className='w-4 h-4 me-2.5 p-0' />
          </Link>
        </li>
        {items.map((item) => (
          <li key={item.label || item.id} className='inline-flex items-center'>
            <img src={arrowRight} alt='Arrow Right' className='rtl:rotate-180 w-6 h-6 text-gray-400 mx-1' />
            {item.href ? (
              <Link to={item.href} className='ms-1 text-sm font-medium text-gray-400 hover:text-primary-500 md:ms-2'>
                {item.label}
              </Link>
            ) : (
              <span className='ms-1 text-sm font-medium text-gray-400 md:ms-2'>{item.label}</span>
            )}
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default memo(Breadcrumbs);
