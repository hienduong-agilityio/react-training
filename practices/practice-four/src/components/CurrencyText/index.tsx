export interface ICurrencyTextProps extends React.HTMLAttributes<HTMLParagraphElement> {
  // Currency: Optional estimated cost of the project, displayed as a financial estimate.
  currency?: number;
}

/**
 * CurrencyText component
 *
 * @returns {JSX.Element} - CurrencyItem element
 */
export const CurrencyText = ({ currency = 0, className = '' }: ICurrencyTextProps): JSX.Element => {
  return <p className={`${className}`}>{currency ? `US$ ${currency}k` : '-'}</p>;
};
