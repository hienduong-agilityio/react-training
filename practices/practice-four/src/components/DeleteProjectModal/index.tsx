// Components
import Modal from '@/components/common/Modal';
import Button from '@/components/common/Button';
import Spinner from '@/components/common/Spinner';

// Enums
import { BUTTON_COLORS, BUTTON_VARIANTS } from '@/enums/theme';

// Hooks
import { useProject } from '@/hooks/useProject';

// Store
import { ToastStore } from '@/stores/toastStore';

// Types
import type { IProjectItemProps } from '@/interfaces';

export interface IDeleteProjectModalProps {
  // projectId: The project id.
  projectId: string;
  // isModalOpen: A boolean indicating whether the modal is open.
  isModalOpen: boolean;
  // onCloseModal: A function to be called when the modal is requested to be closed.
  onCloseModal: () => void;
}

/**
 * The custom DeleteProjectModal component
 *
 * @returns {JSX.Element} - The DeleteProjectModal element
 */
export const DeleteProjectModal = ({
  projectId,
  isModalOpen = false,
  onCloseModal = () => {}
}: IDeleteProjectModalProps): JSX.Element => {
  const project = useProject({ id: projectId || '' });
  const { showToast } = ToastStore();

  const {
    mutate: { deleteProject },
    data: projectDetail,
    isMutating
  } = project;

  const projectName = (projectDetail as IProjectItemProps)?.projectName || 'this project';

  const handleDeleteProjectSuccess = () => {
    showToast('Project Deleted successfully', 'success');
    onCloseModal();
  };

  const handleDeleteProjectError = () => {
    showToast('Failed to deleted project', 'error');
    onCloseModal();
  };

  const handleDelete = () => {
    if (projectId) {
      deleteProject(projectId, {
        onSuccess: handleDeleteProjectSuccess,
        onError: handleDeleteProjectError
      });
    }
  };

  return (
    <Modal isOpen={isModalOpen} onClose={onCloseModal}>
      <div className='bg-light rounded-t-xl p-4 w-[440px]'>
        <p className='text-lg font-semibold text-gray-900 mb-3'>Delete Project</p>
        <p className='leading-6 text-gray-600'>
          Are you sure you want to delete <strong>{projectName}</strong>? If you delete, it will be permanently lost.
        </p>
      </div>
      <div className='bg-light rounded-b-xl flex justify-end gap-4 border-t-2 py-3'>
        <Button onClick={onCloseModal} variant={BUTTON_VARIANTS.OUTLINED}>
          Cancel
        </Button>
        <Button
          onClick={handleDelete}
          disabled={isMutating}
          variant={BUTTON_VARIANTS.CONTAINED}
          color={BUTTON_COLORS.DANGER}
          customClasses='mr-5'
        >
          {isMutating ? <Spinner /> : 'Delete'}
        </Button>
      </div>
    </Modal>
  );
};
