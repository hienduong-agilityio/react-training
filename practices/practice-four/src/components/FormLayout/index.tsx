interface FormLayoutProps {
  title: string;
  subtitle?: string;
  children: React.ReactNode;
}

export const FormLayout = ({ title, subtitle, children }: FormLayoutProps) => (
  <section>
    <div className='flex rounded-t-xl items-center bg-white justify-between p-4'>
      <p className='text-xl font-semibold text-gray-900'>{title}</p>
    </div>
    {subtitle && <p className='text-gray-600 mt-1 mb-4'>{subtitle}</p>}
    {children}
  </section>
);
