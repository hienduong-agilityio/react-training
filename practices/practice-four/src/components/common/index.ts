export { default as Avatar } from './Avatar';
export { default as Button } from './Button';
export { default as InputField } from './InputField';
export { default as Modal } from './Modal';
export { default as Pagination } from './Pagination';
export { default as SelectField } from './SelectField';
export * from './SkeletonLoader';
export { default as Spinner } from './Spinner';
export { default as StatusLabel } from './StatusLabel';
export { default as Tag } from './Tag';
export { default as Toast } from './ToastMessage';
