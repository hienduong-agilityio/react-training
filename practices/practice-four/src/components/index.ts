export { default as Breadcrumbs } from './Breadcrumbs';
export * from './CurrencyText';
export * from './DeleteProjectModal';
export * from './ErrorBoundary';
export * from './ErrorDisplay';
export * from './ProjectDetailSkeleton';
export { default as ProjectFilterDropdown } from './ProjectFilterDropdown';
export * from './ProjectFilterDropdown';
export { default as ProjectItem } from './ProjectItem';
export * from './ProjectItem';
export * from './ProjectOptionsDropdown';
export * from './ProjectTable';
export { default as ResourceTag } from './ResourcesTag';
export { default as SearchBox } from './SearchBox';
export * from './SuspenseLoader';
export { default as Timeline } from './Timeline';
export * from './ToastContainer';
export * from './FormLayout';
export * from './ProjectFormFields';
export * from './SubmitForm';

// Common
export * from './common';
