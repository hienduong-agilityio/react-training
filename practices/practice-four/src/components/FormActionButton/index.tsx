// Libraries
import { useFormStatus } from 'react-dom';

// Components
import { Button, Spinner } from '@/components';

// Enums
import { BUTTON_COLORS, BUTTON_SIZES, BUTTON_VARIANTS } from '@/enums';

interface IFormActionButtonProps {
  children: React.ReactNode;
  type?: 'submit' | 'button';
  color?: BUTTON_COLORS;
  variant?: BUTTON_VARIANTS;
  size?: BUTTON_SIZES;
  disabledOnPending?: boolean;
  showSpinner?: boolean;
  onClick?: () => void;
  customClass?: string;
}

export const FormActionButton = ({
  children,
  type = 'submit',
  variant = BUTTON_VARIANTS.CONTAINED,
  size = BUTTON_SIZES.LARGE,
  color = BUTTON_COLORS.PRIMARY,
  disabledOnPending = true,
  showSpinner = true,
  customClass = '',
  onClick
}: IFormActionButtonProps) => {
  const { pending } = useFormStatus();

  return (
    <Button
      type={type}
      disabled={disabledOnPending && pending}
      variant={variant}
      size={size}
      color={color}
      onClick={onClick}
      customClasses={customClass}
    >
      {pending && showSpinner ? <Spinner /> : children}
    </Button>
  );
};
