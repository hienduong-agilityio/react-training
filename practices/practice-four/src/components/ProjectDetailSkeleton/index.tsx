import { SkeletonLoader } from '@/components';

/**
 * ProjectDetailSkeleton components
 */
export const ProjectDetailSkeleton = () => {
  return (
    <div className='flex items-center p-8 h-full gap-36 space-y-6 flex-row'>
      <div className='flex-shrink-0 w-5/12 h-full flex flex-col items-center'>
        {/* Placeholder for Avatar */}
        <SkeletonLoader width='75%' height='300px' customClassName='mb-4' />
        <SkeletonLoader width='50%' height='20px' />
      </div>

      <div className='flex-grow lg:w-7/12 h-full space-y-6'>
        <SkeletonLoader width='70%' height='30px' customClassName='mb-4' />
        <SkeletonLoader width='20%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='30px' customClassName='mb-4 rounded-md' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='60%' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='60%' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='20px' />
      </div>
    </div>
  );
};
