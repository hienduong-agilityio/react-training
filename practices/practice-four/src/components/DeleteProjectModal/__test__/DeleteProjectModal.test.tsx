// Libraries
import { render, screen, fireEvent } from '@testing-library/react';

// Components
import { DeleteProjectModal, IDeleteProjectModalProps } from '@/components';

// Hooks
import { useProject } from '@/hooks';

const showToastMock = jest.fn();
const hideToastMock = jest.fn();

jest.mock('@/stores/toastStore', () => ({
  ToastStore: () => ({
    showToast: showToastMock,
    hideToast: hideToastMock,
    message: '',
    type: 'success',
    isVisible: false,
    undoEnabled: false,
    timeoutDuration: 3000
  })
}));

// Mock the useProject hook
jest.mock('@/hooks/useProject', () => ({
  useProject: jest.fn()
}));

const deleteProjectMock = jest.fn();

// Define the mock implementation of useProject
(useProject as jest.Mock).mockImplementation(() => ({
  mutate: { deleteProject: deleteProjectMock },
  data: { projectName: 'Test Project' },
  isMutating: false
}));

/*
DeleteProjectModal
  √ renders modal with correct project name
  √ calls onCloseModal when cancel button is clicked
  √ does not render the modal when isModalOpen is false
  √ does not throw error when onCloseModal is not provided
  √ renders "this project" when projectName is not available
  √ calls deleteProject when the delete button is clicked
  √ shows success toast and closes modal on successful deletion
  √ shows error toast and closes modal on failed deletion
  √ shows a spinner while deleting
  √ shows the delete text when not deleting
  √ matches the snapshot
*/
describe('DeleteProjectModal', () => {
  let onCloseMock: jest.Mock;
  let renderDeleteModal: (props?: Partial<IDeleteProjectModalProps>) => ReturnType<typeof render>;

  beforeEach(() => {
    jest.clearAllMocks();
    onCloseMock = jest.fn();

    // Initialize the render function with full props from IDeleteProjectModalProps
    renderDeleteModal = (props: Partial<IDeleteProjectModalProps> = {}) => {
      const defaultProps: IDeleteProjectModalProps = {
        projectId: '123',
        isModalOpen: true,
        onCloseModal: onCloseMock,
        ...props
      };
      return render(<DeleteProjectModal {...defaultProps} />);
    };
  });

  it('matches the snapshot', () => {
    const { container } = renderDeleteModal();
    expect(container).toMatchSnapshot();
  });

  it('renders modal with correct project name', () => {
    renderDeleteModal();

    expect(screen.getByText('Delete Project')).toBeInTheDocument();
    expect(screen.getByText(/Are you sure you want to delete/i)).toBeInTheDocument();
    expect(screen.getByText('Test Project')).toBeInTheDocument();
  });
  //
  it('calls onCloseModal when cancel button is clicked', () => {
    renderDeleteModal();

    fireEvent.click(screen.getByText('Cancel'));
    expect(onCloseMock).toHaveBeenCalled();
  });

  it('does not render the modal when isModalOpen is false', () => {
    renderDeleteModal({ isModalOpen: false });

    expect(screen.queryByText('Delete Project')).not.toBeInTheDocument();
  });

  it('does not throw error when onCloseModal is not provided', () => {
    renderDeleteModal({ onCloseModal: undefined });

    fireEvent.click(screen.getByText('Cancel'));

    expect(screen.getByText('Cancel')).toBeInTheDocument();
  });

  it('renders "this project" when projectName is not available', () => {
    (useProject as jest.Mock).mockReturnValueOnce({
      mutate: { deleteProject: deleteProjectMock },
      data: {},
      isMutating: false
    });

    renderDeleteModal();

    expect(screen.getByText('this project')).toBeInTheDocument();
  });

  it('calls deleteProject when the delete button is clicked', () => {
    renderDeleteModal();

    fireEvent.click(screen.getByText('Delete'));
    expect(deleteProjectMock).toHaveBeenCalledWith('123', expect.any(Object));
  });

  it('shows success toast and closes modal on successful deletion', () => {
    deleteProjectMock.mockImplementationOnce((_, { onSuccess }) => onSuccess());

    renderDeleteModal();

    fireEvent.click(screen.getByText('Delete'));
    expect(showToastMock).toHaveBeenCalledWith('Project Deleted successfully', 'success');
  });

  it('shows error toast and closes modal on failed deletion', () => {
    deleteProjectMock.mockImplementationOnce((_, { onError }) => onError());

    renderDeleteModal();

    fireEvent.click(screen.getByText('Delete'));
    expect(showToastMock).toHaveBeenCalledWith('Failed to delete project', 'error');
  });

  it('shows a spinner while deleting', () => {
    (useProject as jest.Mock).mockReturnValueOnce({
      mutate: { deleteProject: deleteProjectMock },
      data: { projectName: 'Test Project' },
      isMutating: true
    });

    renderDeleteModal();

    expect(screen.getByRole('progressbar')).toBeInTheDocument();
    expect(screen.queryByText('Delete')).not.toBeInTheDocument();
  });

  it('shows the delete text when not deleting', () => {
    (useProject as jest.Mock).mockReturnValueOnce({
      mutate: { deleteProject: deleteProjectMock },
      data: { projectName: 'Test Project' },
      isMutating: false
    });

    renderDeleteModal();

    expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    expect(screen.getByText('Delete')).toBeInTheDocument();
  });
});
