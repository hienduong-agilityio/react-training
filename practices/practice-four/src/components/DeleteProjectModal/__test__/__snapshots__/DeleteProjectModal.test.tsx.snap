// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteProjectModal matches the snapshot 1`] = `
<div>
  <div
    class="fixed inset-0 z-50 flex items-center justify-center w-full h-full bg-gray-800 bg-opacity-60"
  >
    <div
      class="relative rounded-2xl max-w-2xl bg-gray-50 shadow-lg"
      role="dialog"
    >
      <div
        class="bg-light rounded-t-xl p-4 w-[440px]"
      >
        <p
          class="text-lg font-semibold text-gray-900 mb-3"
        >
          Delete Project
        </p>
        <p
          class="leading-6 text-gray-600"
        >
          Are you sure you want to delete 
          <strong>
            Test Project
          </strong>
          ? If you delete, it will be permanently lost.
        </p>
      </div>
      <div
        class="bg-light rounded-b-xl flex justify-end gap-4 border-t-2 py-3"
      >
        <button
          class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 bg-white border-2 text-black px-4 py-2 text-base text-black"
        >
          Cancel
        </button>
        <button
          class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 border-0 text-white px-4 py-2 text-base bg-danger-400 border hover:bg-danger-500 focus:bg-danger-400 hover:text-light focus:text-light mr-5"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
